### PR TITLE
Elin robo class

### PR DESCRIPTION
**Pull request by elinliu2**

### CODE REVIEW

File: .vscode/settings.json
Comment (reubenkw) [inline .vscode/settings.json:1]: editor settings or generated files shouldn't be logged in git. I'd recommend blindly appending [this ](https://gist.github.com/Yousha/3830712334ac30a90eb6041b932b68d7) to the gitignore and remove the bad files

File: .vscode/tasks.json
Comment (reubenkw) [inline .vscode/tasks.json:1]: also remove

File: MotorController.exe
Comment (reubenkw) [inline MotorController.exe:1]: non-source files also shouldn't be tracked in git. The gitignore addition will exclude .exe files. Just remove

File: PID_Ctrl_Test.exe
Comment (reubenkw) [inline PID_Ctrl_Test.exe:1]: also remove.

File: Robot.exe
Comment (reubenkw) [inline Robot.exe:1]: also remove.

File: PID_Ctrl_Test.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#include <iostream>
+#include <fstream>
+
+#include "PID_Ctrl.h"
+
+int main(){
+    PID_Ctrl pid_ctrl(10, 1, 2, 5);
+    
+    std::ofstream myfile;
+    myfile.open ("testing.csv");
```
Comment (reubenkw) [inline PID_Ctrl_Test.cpp]: now is not the time to start reorganizing all these different programs. But at least give the file a slightly better name. Maybe `test_pid_ctrl.csv`

File: PID_Ctrl.cpp
Comment (reubenkw) [inline PID_Ctrl.cpp:1]: I don't really care what convention is picked but let's at least be consistent with naming. For the sake of suggesting something: snake case for file names. ie this should be changed to `pid_ctrl.cpp`

File: Robot.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+double calculate_radius(double delta_x, double delta_y){
+    double L = sqrt(delta_x*delta_x+delta_y*delta_y);
+
+    return L*L/2*std::abs(delta_y);
+}
+
+// From MTE 544 Modeling III IV Lecture Slide 16
+// Pass in -w for left wheel
+double Robot::calculate_wheel_speed(double v, double w){
+    return 1/wheelRadius* ( v + robotWidth * w / 2 + std::pow(robotLength  * w / 2 , 2) / (v + robotWidth * w / 2) );
```
Comment (reubenkw) [inline Robot.cpp]: for the sake of consistency, use the vscode formatter on all files. Either `ctrl` + `alt` + `f` when the cursor is in each file or https://stackoverflow.com/a/52721592. For python, I always use "black"

File: Robot.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+void Robot::driveRobotForward(Point idealPos){
+    Point delta = idealPos - robotPosition;
+
+    while(delta.mag() < robotPosTol){
+        
+        double r = calculate_radius(delta.x, delta.y);
+
+        // TODO: how to figure ideal v?
+        double v = 1;
```
Comment (reubenkw) [inline Robot.cpp]: maybe lets not start with 1 m/s. regardless of what the ideal v is

File: Robot.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+// From MTE 544 Control Lecture Slide 9
+double calculate_radius(double delta_x, double delta_y){
+    double L = sqrt(delta_x*delta_x+delta_y*delta_y);
+
+    return L*L/2*std::abs(delta_y);
+}
+
+// From MTE 544 Modeling III IV Lecture Slide 16
+// Pass in -w for left wheel
```
Comment (reubenkw) [inline Robot.cpp:46]: this assumes our motor signals have a CCW conversion, right? Should note that somewhere

File: Robot.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+Point Robot::getRobotPosition(){
+    return robotPosition;
+}
+
+Point Robot::getArmPosition(){
+    return armPosition;
+}
+
+// how are we going to do this? 
+// should it just be called at the end of each robot move function?
```
Comment (reubenkw) [inline Robot.cpp:32]: Not a change request. Do we want separate threads? Might be good for controllers to have more consistent timings.

File: Robot.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
+#include "Robot.h"
+
+Robot::Robot(double robotLength, double robotWidth, double wheelRadius, Camera & camera, double robotPosTol, double armPosTol) 
+    : robotLength(robotLength), robotWidth(robotWidth), wheelRadius(wheelRadius), camera(camera), robotPosTol(robotPosTol), armPosTol(armPosTol) {
+    // pid terms need to be determined experimentally
```
Comment (reubenkw) [inline Robot.cpp]:
```suggestion
    // TODO: pid terms need to be determined experimentally
```

File: Robot.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        double idealSpeedz = 1; 
+        // TODO: do we need PID controllers ideal speed of servo arm position? 
+        servoArm[x].setIdealSpeed(idealSpeedx);
+        servoArm[y].setIdealSpeed(idealSpeedy);
+        servoArm[z].setIdealSpeed(idealSpeedz);
+
+        updateArmPosition();
+
+        delta = idealPos - armPosition;
+    }
```
Comment (reubenkw) [inline Robot.cpp]: This is a horrible controller. Assuming some very ideal things, requesting a move from (0, 0, 0) to (0, 0, 1) would result in a move to (1, 1, 1).

Missing resetting the ideal speeds to 0 after the loop.

**Review by reubenkw — CHANGES_REQUESTED** [review]

File: Robot.cpp (hunk 4 above)
Comment (elinliu2) [reply Robot.cpp:46]: just added a comment for this here, unless you think the comment should be somewhere else

File: Robot.cpp (hunk 7 above)
Comment (elinliu2) [reply Robot.cpp]: hahahahahahahahaha. good point

File: Robot.cpp (hunk 5 above)
Comment (elinliu2) [reply Robot.cpp:32]: Don't think we want separate threads at the moment. Just want to be able to clear controller.

File: Robot.cpp (hunk 3 above)
Comment (elinliu2) [reply Robot.cpp]: Added a constant at top of file for now at 0.1

File: Robot.cpp (hunk 7 above)
Comment (elinliu2) [reply Robot.cpp]: Replaced with a move servo arm function. Might want to have a separate thread for when we move x and y together @reubenkw

File: Robot.cpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+		drive[frontLeft].setIdealSpeed(leftWheelSpeed);
+		drive[backLeft].setIdealSpeed(leftWheelSpeed);
+		drive[frontRight].setIdealSpeed(rightWheelSpeed);
+		drive[backRight].setIdealSpeed(rightWheelSpeed);
+
+		updateRobotPosition();
+
+		delta = idealPos - robotPosition;
+	}
```
Comment (reubenkw) [inline Robot.cpp:78]: should reset ideal speeds to 0. Even though updateRobotPosition isn't called outside of here.

File: Robot.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	double delta = pos - armPosition[pos];
+	while (delta < armPosTol) {
+		// TODO: how to figure ideal v?
+		double idealSpeed = IDEAL_LINEAR_SPEED;
+
+		// TODO: do we need PID controllers ideal speed of servo arm position? 
+		servoArm[motor].setIdealSpeed(idealSpeed);
+		updateArmPosition();
+		delta = pos - armPosition[pos];
+	}
```
Comment (reubenkw) [inline Robot.cpp:99]: Same as above.

File: main.cpp [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+	Mat image;
+	image = imread("C:\\Users\\elinl\\OneDrive\\Desktop\\School\\capstone\\lena.png");
+	if (!image.data)
+	{
+		printf("No image data \n");
+		return -1;
+	}
+	namedWindow("Display Image", WINDOW_AUTOSIZE);
+	imshow("Display Image", image);
```
Comment (reubenkw) [inline main.cpp]: Remove or comment out this code if the image has been removed?

**Review by reubenkw — CHANGES_REQUESTED** [review]

File: Robot.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		double w = v / r;
+		double leftWheelSpeed = calculate_wheel_speed(v, -1 * w);
+		double rightWheelSpeed = calculate_wheel_speed(v, w);
+
+		drive[frontLeft].setIdealSpeed(leftWheelSpeed);
+		drive[backLeft].setIdealSpeed(leftWheelSpeed);
+		drive[frontRight].setIdealSpeed(rightWheelSpeed);
+		drive[backRight].setIdealSpeed(rightWheelSpeed);
+
+		updateRobotPosition();
```
Comment (elinliu2) [inline Robot.cpp:75]: call output motor voltage to run pid controller

**Review by reubenkw — APPROVED** [review]

**Pull request merged**